bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/TableOfContentsFilter.java

```diff
@@ -0,0 +1,295 @@
+ * store the response content.
+ * Gets the response content from this wrapper, modifies it and copies it into the original response object.
+ */
+@Component(
+    service = Filter.class,
+    property = {Constants.SERVICE_RANKING + "Integer=999"}
+)
+@SlingServletFilter(
+    scope = {SlingServletFilterScope.REQUEST},
+    pattern = "/content/.*",
+    resourceTypes = "cq:Page",
+    extensions = {"html"},
+    methods = {"GET"}
+)
+public class TableOfContentsFilter implements Filter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TableOfContentsFilter.class);
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+
+    }
+
+    @Override
+    public void destroy() {
+
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+        throws IOException, ServletException {
+
+        CharResponseWrapper responseWrapper = new CharResponseWrapper((HttpServletResponseWrapper) response);
+        chain.doFilter(request, responseWrapper);
+        String originalContent = responseWrapper.toString();
+        Boolean containsTableOfContents = (Boolean) request.getAttribute(TableOfContentsImpl.TOC_REQUEST_ATTR_FLAG);
+        if (responseWrapper.getContentType().contains("text/html") &&
+            (containsTableOfContents != null && containsTableOfContents)) {
+
+            Document document = Jsoup.parse(originalContent);
+
+            Elements tocPlaceholderElements = document.getElementsByClass(TableOfContents.TOC_PLACEHOLDER_CLASS);
+            for (Element tocPlaceholderElement : tocPlaceholderElements) {
+                Element tableOfContents = getTableOfContents(tocPlaceholderElement);
+                String id = tocPlaceholderElement.id();
+                tocPlaceholderElement.empty();
+                tocPlaceholderElement.clearAttributes();
+                tocPlaceholderElement.addClass(TableOfContents.TOC_CONTENT_CLASS);
+                if(!id.isEmpty()) {
+                    tocPlaceholderElement.id(id);
+                }
+                if(tableOfContents != null) {
+                    tocPlaceholderElement.appendChild(tableOfContents);
+                    WCMMode wcmMode = WCMMode.fromRequest(request);
+                    if(wcmMode == WCMMode.EDIT || wcmMode == WCMMode.PREVIEW) {
+                        Elements tocTemplatePlaceholderElement = tocPlaceholderElement
+                            .parent()
+                            .select("." + TableOfContents.TOC_TEMPLATE_PLACEHOLDER_CLASS);
+                        tocTemplatePlaceholderElement.remove();
+                    }
+                }
+            }
+
+            CharArrayWriter charWriter = new CharArrayWriter();
+            charWriter.write(document.outerHtml());
+            String alteredContent = charWriter.toString();
+
+            response.setContentLength(alteredContent.length());
```

Comment:
IMHO setting the content length of the request header should not be done in a filter: if different filters set the content length based on their internal logic, it might lead to inconsistencies. It may be the root cause of the sporadic issue [0] I observed while testing the TOC component.

[0] https://github.com/adobe/aem-core-wcm-components/pull/2030#issuecomment-1073706066